hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueContainer.java

```diff
@@ -1216,4 +1217,13 @@ public int getClassId() {
     void setId(long itemId) {
         idGenerator = Math.max(itemId + 1, idGenerator);
     }
+
+    public void prepareForReplication() {
+        // Nullify backup, this is to rescue from the
+        // scenario, in which this partition successively
+        // changes ownership between primary and backup.
+        if (!getItemQueue().isEmpty() && MapUtil.isNullOrEmpty(backupMap)) {
```

Comment:
If understand the internals correctly, it should be impossible to have non-null `backupMap` if `itemQueue` is not null. Probably the actual root cause is that the container isn't converted to the appropriate representation on promotions and demotions, just guessing.

Reply:
Am i not converting it to the appropriate representation with this change?

Reply:
Yes, you are. But the conversion is delayed in time from the actual promotion/demotion and that might be error-prone as this issue demonstrates.

Just out of curiosity, what's the point in keeping an empty `backupMap` while converting the representation in `createPriorityQueue`/`createLinkedList`? Does empty `backupMap` have some special meaning while non-null `itemQueue` co-exists with it?

Reply:
I have updated PR to make the fix more in-place. These was a missing condition to initialize `backupMap` properly and this was manifesting itself after ownership changes. In `getBackup` method, `null backupMap` and `empty backupMap` have different meanings, now i'm initializing `backupMap` in that method.